### PR TITLE
Fix Typescript generic typing for datasource contexts

### PR DESCRIPTION
**Pull request by Tybot204**
Generic typing for datasources in `@apollo/gateway` should be done on a class level so every function / variable within the source uses the same context type. I set a default of `any` on the datasource generics to keep backward compatibility and not require a type or interface to be given to the datasource.

Closes issue #3307.

An example of using this in practice would look like:
```typescript
interface IContext {
  user?: {
    id: string;
  };
}

const gateway = new ApolloGateway({
  buildService({ url }) {
    return new RemoteGraphQLDataSource<IContext>({
      url,
      willSendRequest({ request, context }) {
        // "context" will be typed to IContext now
      }
    });
  }
});
```

### CODE REVIEW

**Comment by abernix** [comment]
Really appreciate the PR!  @trevor-scheer and I have certainly come across this as well and were trying to unwind the motivation for doing it as such originally since that reasoning doesn't seem to be captured in the commit that introduced it.

e.g., Perhaps there is/was a consideration to be had for different requests to different services with different contexts but sharing a common `GraphQLDataSource`?  Thoughts/musings certainly appreciated.

File: packages/apollo-gateway/src/datasources/LocalGraphQLDataSource.ts [hunk 1]
```diff
@@ -9,14 +9,14 @@ import {
 import { enableGraphQLExtensions } from 'graphql-extensions';
 import { GraphQLDataSource } from './types';
 
-export class LocalGraphQLDataSource implements GraphQLDataSource {
+export class LocalGraphQLDataSource<TContext = any> implements GraphQLDataSource<TContext> {
```
Comment (trevor-scheer) [inline packages/apollo-gateway/src/datasources/LocalGraphQLDataSource.ts]: Context shouldn't ever actually be `any` - it _will_ be treated as an object. In other places, we've been going with:
```suggestion
export class LocalGraphQLDataSource<TContext extends Record<string, any> = Record<string, any>> implements GraphQLDataSource<TContext> {
```

File: packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts [hunk 2]
```diff
@@ -19,11 +19,11 @@ import { isObject } from '../utilities/predicates';
 import { GraphQLDataSource } from './types';
 import createSHA from 'apollo-server-core/dist/utils/createSHA';
 
-export class RemoteGraphQLDataSource implements GraphQLDataSource {
+export class RemoteGraphQLDataSource<TContext = any> implements GraphQLDataSource<TContext> {
```
Comment (trevor-scheer) [inline packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts]: Same as above:
```suggestion
export class RemoteGraphQLDataSource<TContext extends Record<string, any> = Record<string, any>> implements GraphQLDataSource<TContext> {
```

File: packages/apollo-gateway/src/datasources/types.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { GraphQLResponse, GraphQLRequestContext } from 'apollo-server-types';
 
-export interface GraphQLDataSource {
-  process<TContext>(
+export interface GraphQLDataSource<TContext = any> {
```
Comment (trevor-scheer) [inline packages/apollo-gateway/src/datasources/types.ts]: Same as above:
```suggestion
export interface GraphQLDataSource<TContext extends Record<string, any> = Record<string, any>> {
```

**Comment by Tybot204** [comment]
@abernix @trevor-scheer In regards to allowing different context types depending on the underlying service, maybe this can still be accomplished using a context type at the class level. This is just me spitballing so I haven't put much thought behind it, but if someone needs different contexts they could create their own datasource class for their service and extend `RemoteGraphQLDataSource`, defining the new context type however needed.

**Comment by trevor-scheer** [comment]
Hey @Tybot204, this is ready to go. I'm having a hard time pushing up a changelog entry to your branch. Here's a patch if you'd like to apply it via `git apply my-patch.patch`, or feel free to write your own. Thanks again!

```
commit e4ba9a0c0bb8379bdc1572fa9652291c58c54590
Author: Trevor Scheer <trevor.scheer@gmail.com>
Date:   Fri Mar 20 13:57:53 2020 -0700

    Update changelog

diff --git a/packages/apollo-gateway/CHANGELOG.md b/packages/apollo-gateway/CHANGELOG.md
index 56353573..43d15532 100644
--- a/packages/apollo-gateway/CHANGELOG.md
+++ b/packages/apollo-gateway/CHANGELOG.md
@@ -1,5 +1,9 @@
 # CHANGELOG for `@apollo/gateway`
 
+## vNEXT
+
+- Fix Typescript generic typing for datasource contexts [#3865](https://github.com/apollographql/apollo-server/pull/3865) This is a fix for the `TContext` typings of the gateway's exposed `GraphQLDataSource` implementations. In their current form, they don't work as intended, or in any manner that's useful for typing the `context` property throughout the class methods. This introduces a type argument `TContext` to the class itself (which defaults to `Record<string, any>` for existing implementations) and removes the non-operational type arguments on the class methods themselves.
+
 ## 0.13.2
 
 - __BREAKING__: The behavior and signature of `RemoteGraphQLDataSource`'s `didReceiveResponse` method has been changed.  No changes are necessary _unless_ your implementation has overridden the default behavior of this method by either extending the class and overriding the method or by providing `didReceiveResponse` as a parameter to the `RemoteGraphQLDataSource`'s constructor options.  Implementations which have provided their own `didReceiveResponse` using either of these methods should view the PR linked here for details on what has changed.  [PR #3743](https://github.com/apollographql/apollo-server/pull/3743)

```

**Comment by trevor-scheer** [comment]
Thanks so much @Tybot204!

**Pull request merged**